### PR TITLE
Update to Arrow 35.0.0

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change

Update to Arrow 35.0.0

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -2979,10 +2979,6 @@ mod tests {
             ScalarValue::Decimal128(None, 10, 2),
             ScalarValue::try_from_array(&array, 3).unwrap()
         );
-        assert_eq!(
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:2982]: This test was wrong, index 4 is beyond the bounds of the array

File: datafusion/core/tests/dataframe_functions.rs [hunk 2]
```diff
@@ -398,10 +398,10 @@ async fn test_fn_regexp_match() -> Result<()> {
         "+-----------------------------------+",
         "| regexpmatch(test.a,Utf8(\"[a-z]\")) |",
         "+-----------------------------------+",
-        "| []                                |",
-        "| []                                |",
-        "| []                                |",
-        "| []                                |",
+        "| [a]                               |",
```
Comment (alamb) [inline datafusion/core/tests/dataframe_functions.rs:401]: 🎉

File: datafusion/core/tests/sql/timestamp.rs [hunk 3]
```diff
@@ -1411,14 +1411,14 @@ async fn cast_timestamp_before_1970() -> Result<()> {
 
     assert_batches_eq!(expected, &actual);
 
-    let sql = "select cast('1969-01-01T00:00:00.1Z' as timestamp);";
+    let sql = "select cast('1969-01-01T00:00:00.100Z' as timestamp);";
```
Comment (alamb) [inline datafusion/core/tests/sql/timestamp.rs:1414]: what happens without this change?

Comment (tustvold) [reply datafusion/core/tests/sql/timestamp.rs:1414]: An error, it was a mis-reading of the RFC spec, I thought it restricted to multiples of 3 for subsecond precision, I'm working on a fix

Comment (alamb) [reply datafusion/core/tests/sql/timestamp.rs:1414]: https://github.com/apache/arrow-rs/pull/3858 I think is related

Comment (alamb) [reply datafusion/core/tests/sql/timestamp.rs:1414]: I would tend to think this regression should block the upgrade of datafusion, but would defer to others. 

maybe we can mark this PR as "ready for review" to get some other opinions 🤔

File: datafusion/core/tests/sql/set_variable.rs [hunk 4]
```diff
@@ -414,7 +414,7 @@ async fn set_time_zone_bad_time_zone_format() {
             .await
             .unwrap();
     let err = pretty_format_batches(&result).err().unwrap().to_string();
-    assert_eq!(err, "Parser error: Invalid timezone \"+08:00:00\": Expected format [+-]XX:XX, [+-]XX, or [+-]XXXX");
+    assert_eq!(err, "Parser error: Invalid timezone \"+08:00:00\": only offset based timezones supported without chrono-tz feature");
```
Comment (comphead) [inline datafusion/core/tests/sql/set_variable.rs:417]: perhaps we can improve the error message a bit in a future :blush:

File: datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs [hunk 5]
```diff
@@ -431,7 +431,7 @@ mod tests {
             .project(proj)?
             .build()?;
 
-        let expected = "Error parsing 'I'M NOT A TIMESTAMP' as timestamp";
+        let expected = "Error parsing timestamp from 'I'M NOT A TIMESTAMP'";
```
Comment (comphead) [inline datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs:434]: 👍

**Comment by comphead** [comment]
Regression issue here preventing tests to run
`select '2020-01-02 01:01:11.1234567890Z'::timestamp` now fails
but `select '2020-01-02 01:01:11.123456789Z'::timestamp` is ok

@alamb @tustvold

**Comment by comphead** [comment]
> > Regression issue here preventing tests to run
> 
> Thanks @comphead -- I think that is [apache/arrow-rs#3859](https://github.com/apache/arrow-rs/issues/3859)
> 
> There were a few regressions, in arrow 35.0.0 so I think we should probably skip arrow 35.0.0 and wait for arrow 36.0.0: [apache/arrow-rs#3889](https://github.com/apache/arrow-rs/issues/3889)

Do we have a list of regression to be fixed in arrow 36.0.0?

**Comment by tustvold** [comment]
Closing in favor of https://github.com/apache/arrow-datafusion/pull/5685

**Pull request closed without merging**